### PR TITLE
Feature/reader

### PR DESCRIPTION
**Pull request by paulo-haas**
Our team's task is to convert Java files into a generic data structure.

To parse Java files we're using are using a [library](http://javaparser.org/).  We added the library's archive file to the repository and included it in the project.

In the first step we implemented a Java reader (cf. `JavaReader.java`) and included it to the plugin file making it usable.  To debug the result visually we are using a generator (cf. `TreeConverter.java`), which returns a code in the [DOT graph description language](https://graphviz.org/doc/info/lang.html).

In the Java reader a Java file is parsed by the library which returns an abstract syntax tree which is converted into the generic data structure by our Visitor (cf. `Visitor.java`).  The generic data structure is a tree using the framework's implementations of nodes and trees.  The visitor starts at the root node - the compilation unit - and traverses each node's childs.  In every step our Visitor is step-by-step building up the generic data structure tree by adding new nodes or attributes.  To avoid hardcoded strings we are using the names of the library or our expansion (cf. `JavaNodeTypes.java`), which is an enum introducing more options to fit our needs.  The high variety of Java statements, expressions, types etc. make the visitor such a large file.

A lot of operations are repeated at different points of our Visitor so we introduced a utility class (cf. `VisitorUtil.java`).  Currently we're working on improving the tree and correcting the last errors, thus a lot of it is stil WIP and utilization of the helper class is kind of low.  

The writer is fairly new and not fully developed.  There will be heavy changes in the next pull requests.

### CODE REVIEW

**Comment by paulo-haas** [comment]
I've split up the internal structure (`JavaNodeTypes` -> `JavaNodeTypes` and `JavaNodeTypes`) and increased the usage of our utility class (`VisitorUtil`).  Besides there've been other small changes.

File: de.tu_bs.cs.isf.e4cf.compare.data_structures/src/de/tu_bs/cs/isf/e4cf/compare/data_structures/io/reader/JavaReader.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+/***
+ * 
+ * @author Serkan Acar
+ * @author Hassan Smaoui
+ * @author Pascal Blum
+ * @author Paulo Haas
+ *
+ */
+
```
Comment (KamilRosiak) [inline de.tu_bs.cs.isf.e4cf.compare.data_structures/src/de/tu_bs/cs/isf/e4cf/compare/data_structures/io/reader/JavaReader.java]: JavaDoc

File: de.tu_bs.cs.isf.e4cf.compare.data_structures/src/de/tu_bs/cs/isf/e4cf/compare/data_structures/io/reader/JavaReader.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+		if (isFileSupported(element)) {
+			String s = FileStreamUtil.readLineByLine(Paths.get(element.getAbsolutePath()));
+			String fileName = Paths.get(element.getAbsolutePath()).getFileName().toString();
+			CompilationUnit cu = StaticJavaParser.parse(s);
+
+			Node rootNode = new NodeImpl(fileName);
+			
+			Visitor visitor = new Visitor();
+
```
Comment (KamilRosiak) [inline de.tu_bs.cs.isf.e4cf.compare.data_structures/src/de/tu_bs/cs/isf/e4cf/compare/data_structures/io/reader/JavaReader.java]: Use the Autoformater of eclipse Ctrl + Shift + F

File: de.tu_bs.cs.isf.e4cf.compare.data_structures/src/de/tu_bs/cs/isf/e4cf/compare/data_structures/io/reader/Visitor.java [hunk 3]
```diff
@@ -0,0 +1,989 @@
...
+/**
+ * Part of the custom visitor class. This class extends VoidVisitorAdapter
+ * (https://www.javadoc.io/doc/com.github.javaparser/javaparser-core/latest/com/github/javaparser/ast/visitor/VoidVisitorAdapter.html).
+ * 
+ * @author Serkan Acar
+ * @author Pascal Blum
+ * @author Paulo Haas
+ * @author Hassan Smaoui
+ *
+ */
```
Comment (KamilRosiak) [inline de.tu_bs.cs.isf.e4cf.compare.data_structures/src/de/tu_bs/cs/isf/e4cf/compare/data_structures/io/reader/Visitor.java]: I think JavaVisitior would be a better name for this class

File: de.tu_bs.cs.isf.e4cf.compare.data_structures/src/de/tu_bs/cs/isf/e4cf/compare/data_structures/io/reader/Visitor.java [hunk 4]
```diff
@@ -0,0 +1,989 @@
...
+public class Visitor extends VoidVisitorAdapter<Node> {
+	/**
```
Comment (KamilRosiak) [inline de.tu_bs.cs.isf.e4cf.compare.data_structures/src/de/tu_bs/cs/isf/e4cf/compare/data_structures/io/reader/Visitor.java]: Are those comments generated? I would expect comments such as, "This method process the compilation unit and is called ...

File: de.tu_bs.cs.isf.e4cf.compare.data_structures/src/de/tu_bs/cs/isf/e4cf/compare/data_structures/io/reader/Visitor.java [hunk 5]
```diff
@@ -0,0 +1,989 @@
...
+	 * https://www.javadoc.io/static/com.github.javaparser/javaparser-core/3.17.0/com/github/javaparser/ast/CompilationUnit.html
+	 */
+	@Override
+	public void visit(CompilationUnit n, Node arg) {
+		Node cu = VisitorUtil.Parent(n, arg);
+
```
Comment (KamilRosiak) [inline de.tu_bs.cs.isf.e4cf.compare.data_structures/src/de/tu_bs/cs/isf/e4cf/compare/data_structures/io/reader/Visitor.java]: Auto - Formatter

File: de.tu_bs.cs.isf.e4cf.compare.data_structures/src/de/tu_bs/cs/isf/e4cf/compare/data_structures/io/reader/Visitor.java [hunk 6]
```diff
@@ -0,0 +1,989 @@
...
+		// Imports
+		Node imports = VisitorUtil.Node(JavaNodeTypes.Import, cu);
+		int importSize = n.getImports().size();
+		VisitorUtil.AddAttribute(imports, JavaAttributesTypes.Children, String.valueOf(importSize));
+		for (int i = 0; i < importSize; i++) {
```
Comment (KamilRosiak) [inline de.tu_bs.cs.isf.e4cf.compare.data_structures/src/de/tu_bs/cs/isf/e4cf/compare/data_structures/io/reader/Visitor.java]: Yeah, you have to write more comment above the method above critical parts such as this :P

File: de.tu_bs.cs.isf.e4cf.compare.data_structures/src/de/tu_bs/cs/isf/e4cf/compare/data_structures/io/reader/VisitorUtil.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+ *
+ */
+
+public class VisitorUtil {
+	/**
+	 * Creates a leaf node
+	 * 
+	 * @param n JavaParser Node
+	 * @param arg Parent Node
+	 */
```
Comment (KamilRosiak) [inline de.tu_bs.cs.isf.e4cf.compare.data_structures/src/de/tu_bs/cs/isf/e4cf/compare/data_structures/io/reader/VisitorUtil.java]: Starting method names with a capital letter is not a Java Conevention. I think its a c# convention.

File: de.tu_bs.cs.isf.e4cf.compare.data_structures/src/de/tu_bs/cs/isf/e4cf/compare/data_structures/io/reader/VisitorUtil.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+	public static Node Leaf(com.github.javaparser.ast.Node n, Node arg) {
```
Comment (KamilRosiak) [inline de.tu_bs.cs.isf.e4cf.compare.data_structures/src/de/tu_bs/cs/isf/e4cf/compare/data_structures/io/reader/VisitorUtil.java]: Method names could be more precise such as createLeaf

**Review by KamilRosiak — COMMENTED** [review]
Auto-Formatter mal benutzen und die comments mal durch gehen sonst sieht es ganz gut aus.

**Pull request closed without merging**